Implement Jacobian criterion for projective varieties

paemurru opened the pull request:
Implement Jacobian criterion for projective varieties

The is_smooth function for projective schemes now has three algorithms for checking smoothness:

  * covered version of the Jacobian criterion - uses decomposition_info, also works for non-projective covered schemes, implemented only for equidimensional schemes of finite type over a field,
  * projective Jacobian criterion - only for equidimensional schemes of finite type over a field,
  * checking that the affine cone is smooth outside the origin - works for arbitrary schemes over arbitrary base schemes.

# Benchmarks

In the benchmarks, the result of `is_equidimensional` seems to be cached. For the correct results, add the time for `is_equidimensional` to both the projective Jacobian criterion and the covered scheme times.

```
julia> using BenchmarkTools

julia> function is_smooth_benchmark(I::MPolyIdeal{<:MPolyDecRingElem{<:FieldElem, <:MPolyRingElem{<:FieldElem}}})
  text_width = length("* is_equidimensional  ")
  print(rpad("* is_equidimensional", text_width))
  @time begin
    X = projective_scheme(I);
    is_equidimensional(X);
  end
  for algo in [
    :projective_jacobian,
    :covered_jacobian,
    :affine_cone,
  ]
    print(rpad(string(algo), text_width))
    @btime begin
      X = projective_scheme(I)
      is_smooth(X; algorithm=$algo)
    end
  end
end;

julia> X = castelnuovo();

julia> I = defining_ideal(X);

julia> is_smooth_benchmark(I);
* is_equidimensional  0.002994 seconds (7.90 k allocations: 230.954 KiB)
projective_jacobian   5.536 ms (164627 allocations: 5.20 MiB)
covered_jacobian      8.625 ms (171538 allocations: 4.59 MiB)
affine_cone           18.678 ms (169989 allocations: 3.74 MiB)

julia> X = rational_d7_pi4();

julia> I = defining_ideal(X);

julia> is_smooth_benchmark(I);
* is_equidimensional  0.041981 seconds (59.36 k allocations: 1.843 MiB)
projective_jacobian   697.084 ms (11760108 allocations: 363.75 MiB)
covered_jacobian      664.238 ms (9772454 allocations: 218.46 MiB)
affine_cone           1.142 s (12452732 allocations: 249.54 MiB)

julia> X = rational_d10_pi9_quart_1();

julia> I = defining_ideal(X);

julia> is_smooth_benchmark(I);
* is_equidimensional 98.823506 seconds (303.26 k allocations: 23.185 MiB)
projective_jacobian   8.689 s (116793197 allocations: 3.81 GiB)
covered_jacobian      6.571 s (94217947 allocations: 2.16 GiB)
affine_cone           116.049 s (123218985 allocations: 3.07 GiB)

julia> A, (x_1, x_2, x_3, x_4) = grade(QQ["x_1", "x_2", "x_3", "x_4"][1]);

julia> complete_intersection_curve_of_deg_16_in_P3 = ideal(A, [
  8*x_1^0*x_2^0*x_3^0*x_4^4 + 10*x_1^0*x_2^0*x_3^1*x_4^3 + 8*x_1^0*x_2^0*x_3^2*x_4^2 + 16*x_1^0*x_2^0*x_3^3*x_4^1 + 9*x_1^0*x_2^0*x_3^4*x_4^0 + 7*x_1^0*x_2^1*x_3^0*x_4^3 + 9*x_1^0*x_2^1*x_3^1*x_4^2 + 4*x_1^0*x_2^1*x_3^2*x_4^1 + 11*x_1^0*x_2^1*x_3^3*x_4^0 + 4*x_1^0*x_2^2*x_3^0*x_4^2 + 14*x_1^0*x_2^2*x_3^1*x_4^1 + 3*x_1^0*x_2^2*x_3^2*x_4^0 + 13*x_1^0*x_2^3*x_3^0*x_4^1 + 17*x_1^0*x_2^3*x_3^1*x_4^0 + 6*x_1^0*x_2^4*x_3^0*x_4^0 + 17*x_1^1*x_2^0*x_3^0*x_4^3 + 14*x_1^1*x_2^0*x_3^1*x_4^2 + 3*x_1^1*x_2^0*x_3^2*x_4^1 + 4*x_1^1*x_2^0*x_3^3*x_4^0 + 11*x_1^1*x_2^1*x_3^0*x_4^2 + 4*x_1^1*x_2^1*x_3^1*x_4^1 + 3*x_1^1*x_2^1*x_3^2*x_4^0 + 1*x_1^1*x_2^2*x_3^0*x_4^1 + 16*x_1^1*x_2^2*x_3^1*x_4^0 + 8*x_1^1*x_2^3*x_3^0*x_4^0 + 11*x_1^2*x_2^0*x_3^0*x_4^2 + 2*x_1^2*x_2^0*x_3^1*x_4^1 + 12*x_1^2*x_2^0*x_3^2*x_4^0 + 6*x_1^2*x_2^1*x_3^0*x_4^1 + 13*x_1^2*x_2^1*x_3^1*x_4^0 + 8*x_1^2*x_2^2*x_3^0*x_4^0 + 16*x_1^3*x_2^0*x_3^0*x_4^1 + 15*x_1^3*x_2^0*x_3^1*x_4^0 + 6*x_1^3*x_2^1*x_3^0*x_4^0 + 3*x_1^4*x_2^0*x_3^0*x_4^0,
  5*x_1^0*x_2^0*x_3^0*x_4^4 + 3*x_1^0*x_2^0*x_3^1*x_4^3 + 10*x_1^0*x_2^0*x_3^2*x_4^2 + 9*x_1^0*x_2^0*x_3^3*x_4^1 + 1*x_1^0*x_2^0*x_3^4*x_4^0 + 14*x_1^0*x_2^1*x_3^0*x_4^3 + 8*x_1^0*x_2^1*x_3^1*x_4^2 + 16*x_1^0*x_2^1*x_3^2*x_4^1 + 14*x_1^0*x_2^1*x_3^3*x_4^0 + 14*x_1^0*x_2^2*x_3^0*x_4^2 + 13*x_1^0*x_2^2*x_3^1*x_4^1 + 0*x_1^0*x_2^2*x_3^2*x_4^0 + 13*x_1^0*x_2^3*x_3^0*x_4^1 + 3*x_1^0*x_2^3*x_3^1*x_4^0 + 10*x_1^0*x_2^4*x_3^0*x_4^0 + 13*x_1^1*x_2^0*x_3^0*x_4^3 + 13*x_1^1*x_2^0*x_3^1*x_4^2 + 11*x_1^1*x_2^0*x_3^2*x_4^1 + 15*x_1^1*x_2^0*x_3^3*x_4^0 + 11*x_1^1*x_2^1*x_3^0*x_4^2 + 10*x_1^1*x_2^1*x_3^1*x_4^1 + 1*x_1^1*x_2^1*x_3^2*x_4^0 + 7*x_1^1*x_2^2*x_3^0*x_4^1 + 7*x_1^1*x_2^2*x_3^1*x_4^0 + 10*x_1^1*x_2^3*x_3^0*x_4^0 + 2*x_1^2*x_2^0*x_3^0*x_4^2 + 4*x_1^2*x_2^0*x_3^1*x_4^1 + 6*x_1^2*x_2^0*x_3^2*x_4^0 + 4*x_1^2*x_2^1*x_3^0*x_4^1 + 4*x_1^2*x_2^1*x_3^1*x_4^0 + 7*x_1^2*x_2^2*x_3^0*x_4^0 + 15*x_1^3*x_2^0*x_3^0*x_4^1 + 16*x_1^3*x_2^0*x_3^1*x_4^0 + 12*x_1^3*x_2^1*x_3^0*x_4^0 + 12*x_1^4*x_2^0*x_3^0*x_4^0
]);

julia> is_smooth_benchmark(complete_intersection_curve_of_deg_16_in_P3);
* is_equidimensional  0.777904 seconds (926.87 k allocations: 61.203 MiB, 99.44% compilation time)
projective_jacobian   9.738 s (116793197 allocations: 3.81 GiB)
covered_jacobian      7.195 s (94217930 allocations: 2.16 GiB)
affine_cone           114.242 s (123218985 allocations: 3.07 GiB)

julia> I = grassmann_pluecker_ideal(2, 5);

julia> S, _ = grade(base_ring(I));

julia> I = ideal(S, S.(gens(I)));

julia> is_smooth_benchmark(I);
* is_equidimensional  0.003757 seconds (2.91 k allocations: 96.445 KiB)
projective_jacobian   17.214 ms (524540 allocations: 22.84 MiB)
covered_jacobian      78.668 ms (1481782 allocations: 84.25 MiB)
affine_cone           172.092 ms (485141 allocations: 24.74 MiB)
```

paemurru committed:
Affine cone algorithm

paemurru committed:
Remove whitespace at end of lines

paemurru committed:
Add Jacobian criterion, tests, documentation

paemurru committed:
Fix typo :jacobian -> :jacobi

paemurru committed:
Show documentation in html file

paemurru committed:
Run bibtool

jankoboehm commented:
This is primarily interesting of couse for codimension > 1. For the Grassmanian we would expect that the covered version outperforms the other two by far. What is called on the affine patches? Is it just the Jacobian criterion or the algorithm from https://doi.org/10.1016/j.jsc.2017.05.001 or alike?

jankoboehm commented:
Also note: This is also primarily interesting for examples with substantially longer run-time.

wdecker commented:
In addition to what @jankoboehm said with respect to  https://doi.org/10.1016/j.jsc.2017.05.001 , and aiming for examples with a longer running time, I suggest that you apply this to some of the hard-coded Nongeneral Type Surfaces in P4 as described in the documention. For example:

`X = castelnuovo();`
`I = defining_ideal(X);`

`X = rational_d7_pi4();`
`I = defining_ideal(X);`

`X = rational_d10_pi9_quart_1();`
`I = defining_ideal(X);`

Here, the method from the above cited paper should outperform everything else.

lgoettgens commented on docs/src/AlgebraicGeometry/Schemes/ProjectiveSchemes.md:
````diff
@@ -85,7 +85,10 @@ To facilitate the interplay between an `AbsProjectiveScheme` and the affine char
-Further properties of projective schemes (all self-explanatory):
+Further properties of projective schemes:
+```@docs
+    is_smooth(P::AbsProjectiveScheme)
+```
 ```
````
```suggestion
```
`is_smooth` is already in the list

lgoettgens commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -16,15 +16,73 @@ end
 
-# TODO: Jacobi criterion
-function is_smooth(P::AbsProjectiveScheme; algorithm=:covered)
+Check whether the scheme `X` is smooth.
```
In the function header a few lines above it is called `P` instead of `X`.

lgoettgens commented on docs/oscar_references.bib:
```diff
@@ -1385,6 +1385,19 @@ @PhDThesis{Lev05
 
+@Book{Liu06,
+  author        = {Liu, Qing},
+  title         = {Algebraic geometry and arithmetic curves. {Transl}. by {Reinie} {Ern{\'e}}},
```
Our bibtex implementation does not like braces and escapes in the title. Please use unicode instead. Casing is preserved without anything to do.

lgoettgens reviewed (commented):
some minor nitpicks

simonbrandhorst commented:
> This is primarily interesting of couse for codimension > 1. For the Grassmanian we would expect that the covered version outperforms the other two by far. What is called on the affine patches? Is it just the Jacobian criterion or the algorithm from https://doi.org/10.1016/j.jsc.2017.05.001 or alike?

@afkafkafk13  will know

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -16,15 +16,73 @@ end
+    elseif algorithm == :affine_cone
+      aff = affine_cone(P)[1]
+      origin = subscheme(aff, coordinates(aff))
+      return singular_locus_reduced(aff)[1] == origin
```
This check uses a radical computation and an equality check. Could you check whether some variant using saturation or radical membership might be more efficient?

paemurru replied:
I benchmarked three approaches on the example `castelnuovo()`:

* `singular_locus_reduced`,
* saturation,
* radical membership.

They all seemed roughly the same speed.

afkafkafk13 commented:
> > This is primarily interesting of couse for codimension > 1. For the Grassmanian we would expect that the covered version outperforms the other two by far. What is called on the affine patches? Is it just the Jacobian criterion or the algorithm from https://doi.org/10.1016/j.jsc.2017.05.001 or alike?
> 
> @afkafkafk13 will know

As of now, is_regular, which implements the algorithm of the article (currently in a sequential way), is not in Master. There is a PR coming up either before or shortly after Xmas. The variant '':covered'' uses something else, which has been implemented by @HechtiDerLachs

HereAround committed:
[FTheoryTools] Add new literature model constructor

paemurru committed:
Merge branch 'oscar-system:master' into EP/JacobiCriterion

paemurru committed:
Fix minor nitpicks by lgoettgens

paemurru committed:
Merge branch 'oscar-system:master' into EP/JacobiCriterion

paemurru committed:
Improve all three smoothness algorithms

paemurru commented:
Edited the pull request:
  * The projective Jacobian criterion and the affine cone algorithms now check that dimension of singular locus is at most 0, improving performance.
  * Covered scheme algorithm now uses `decomposition_info` of @HechtiDerLachs .
  * Added the surfaces of @wdecker to the benchmarks.
  * Added minor fixes by @lgoettgens .

Have not tried radical membership that @afkafkafk13 suggested nor the algorithm in [https://doi.org/10.1016/j.jsc.2017.05.001](https://doi.org/10.1016/j.jsc.2017.05.001) that @jankoboehm  suggested.

fingolfin committed:
Re-export deprecated bindings from AA and Nemo (#3132)

StevellM committed:
resolve CI issues Hecke (#3140)

JohnAAbbott committed:
Fix conversion bug for PBWAlgQuo elements (#3133)

thofma committed:
test: make worker print test name at the beginning (#3142)

ederc committed:
Normal form from msolve (#3116)

* adds AlgebraicSolving-based normal form using f4

* adds tests for internal normal form implementations

* fixes tests for internal normal form implementations

* bumps AlgebraicSolving to v0.4.3 to ensure GB caching in f4 normal form computations

* removes *_assure calls

* removes useless homomorphism in Singular

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/CoveredSchemes/Objects/Properties.jl:
```diff
@@ -32,16 +32,35 @@ Return the boolean value whether a covered scheme `X` is smooth.
 end
 
+# Use decomposition_info
+function _jacobian_criterion(X::CoveredScheme{<:Field})
```
@HechtiDerLachs 
Is a decomposition info guaranteed to exist and to be set?

paemurru replied:
According to its docstring, no.

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/CoveredSchemes/Objects/Properties.jl:
```diff
@@ -32,16 +32,35 @@ Return the boolean value whether a covered scheme `X` is smooth.
 
+# Use decomposition_info
+function _jacobian_criterion(X::CoveredScheme{<:Field})
+  if !isdefined(X, :coverings)
+    return true
+  end
```
what exactly do we do here? If we want to check that the scheme is empty,
then we better do `is_empty(X)`?

simonbrandhorst replied:
O.k. an empty covered scheme may have many empty patches .. but it is non-trivial to check that they are actually empty.

paemurru replied:
https://github.com/oscar-system/Oscar.jl/pull/3111#discussion_r1442826980

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -16,19 +16,104 @@ end
+    if !(algorithm in algorithms)
+      throw(ArgumentError(
+        "the optional argument to the function is_smooth can only be one"
+         + " of the following: " + join(algorithms, ", ") + "."
```
```suggestion
         * " of the following: " * join(algorithms, ", ") * "."
```

paemurru replied:
Thanks, corrected.

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -16,19 +16,104 @@ end
+    if algorithm == :covered_jacobian
+      return _jacobian_criterion(covered_scheme(P))
+    elseif algorithm == :affine_cone
+      aff = affine_cone(P)[1]
```
An equivalent way to write this is
```suggestion
      aff, _ = affine_cone(P)
```
Personally, I have no preference.

paemurru replied:
Edited according to your suggestion.

simonbrandhorst commented:
Looking at your benchmarks, it seems to me that the `:affine_cone` algorithm is consistently slower, even if we include the `is_equidimensional` timing. Thus I would make it a default only if none of the other algorithms applies.

ThomasBreuer committed:
fix `quo` if the result is a group of automorphisms (#3148)

lgoettgens committed:
Resolve some issues of unbound type parameters (#3060)

* Track number of unbound type parameters

thofma committed:
fix: minor fix for upcoming broadcasting changes (#3149)

The upcoming broadcasting changes will preserve the type of the matrix,
e.g., turn Oscar matrices into Oscar matrices and for those, `lcm(...)`
is not implemented.

StevellM committed:
Cleanup of  `SymmetricIntersections` (#3150)

* first round

* final round

paemurru committed:
docstring

paemurru committed:
Merge branch 'master' into EP/JacobiCriterion

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -16,19 +16,113 @@ end
 end
 
+@doc raw"""
+    function is_smooth(P::AbsProjectiveScheme; algorithm=:default) -> Bool
```
```suggestion
    is_smooth(P::AbsProjectiveScheme; algorithm=:default) -> Bool
```

paemurru replied:
Corrected.

paemurru committed:
Affine cone for arbitrary base scheme

paemurru committed:
Fix tests, add Görtz–Wedhorn to bib
